code review #1

tschimbr opened the pull request:
This is a dummy PR in order to be able to comment on changes since the start of development and hence do a code review.

Commit:
Changed semantical_categories from String to Array and corrected typo

Commit:
Added association do category

Commit:
Generated controller worGenerated controller wor

Commit:
Renamed controller, name was just singular

Commit:
Added resources words to routes

Commit:
Had to delete the seed_helper otherwise rake wouldn't work

Commit:
hat to rename folder from word words

Commit:
Commented out gmaps, we don't need at the moment and it throws an error

Commit:
Commented the nav bar out, to see if it works.

Commit:
Implemented view 'new'

Commit:
Edited new and create method.

Commit:
Added a root path

Commit:
Merge branch 'word_controller' into first_view

Commit:
Changed named of the labels

Commit:
Implemented function word_params

Commit:
Commented out dependency and had do change semantical_category to string (just temporary)

Commit:
Deleted blank line

Commit:
Created view info/home and editet routes file:

Commit:
Replaced navbar, with standard navbar from bootstrap

Commit:
invoce the whole navbar, commented out home.scss and deleted blank lines in home.html

Commit:
Editet the words/new page.

Commit:
Deleted footer content and added som lorem.

Commit:
Added sticky footer and overrided background color

Commit:
Added a footer which now appears everywhere.

Commit:
Created responsive table in words/index. Dont care about the infos/home file.

Commit:
Link 'Libary' will go now to words#index

Commit:
Link 'New flash card' works now.

Commit:
Link 'Home' works now.

tschimbr committed:
Merge pull request #1 from eonum/word_model

Word model

tschimbr committed:
Merge pull request #2 from eonum/word_controller

Word controller

Commit:
Merge branch 'first_view'

Commit:
Merge branch 'home_screen'

Commit:
Commit after rails generate model Category

Commit:
Added/commented out associations

Commit:
implemented destroy method in wordsController.

Commit:
Added 'show' button

Commit:
Button works nowButton works now.

Commit:
Set up basic show view, styling will come later on.

Commit:
updated word model, attributes are spilt up in de and en.

Commit:
Replaced the old attribute names with the new one.

Commit:
Added new fields for the different languages.

Commit:
Prouduct of rails generate controller categories

Commit:
Created views/categories/new and the according stuff in controller and routes

Commit:
Created views/categories/index.

Commit:
Created the edit view. Don't get confused by the branch name.

Commit:
Merge branch 'word_model'

Commit:
Merge branch 'category_model'

Commit:
Merge branch 'category_create'

Commit:
Merge branch 'category_show'

Commit:
Merge branch 'category_update'

Commit:
Implemented the update function

Commit:
Added blank line.

Commit:
Merge branch 'category_update'

Commit:
Implemented destroy

Commit:
Had to edit some attribute names in words_controller and words/index

Commit:
Merge branch 'master' into destory_word

Commit:
Implemented destroy for word

Commit:
Implemented Update for words.

Commit:
Changend attributeind words/show.

Commit:
Added some icons to links in navbar.

Commit:
Configured bootstrap_and overrides and application.rb because icons were different. I used this stackoverflow as guide http://stackoverflow.com/questions/21962775/bootstrap-3rails-4-certain-glyphicons-not-working.

Commit:
Added a comment.

Commit:
Added some more icons to the nav barAdded some more.

Commit:
Forgot at some links the :aria => {:hidden => "true"}.

Commit:
implemented drop down.

Commit:
Changed attribute semantical_categories_en type form String to Array.

Commit:
put the resources in routes into the scope loop

Commit:
Deleted row in application.html.hamel the sites will handel it

Commit:
Nested the content different

Commit:
Added new colos for coloring the page

Commit:
Added spacer to footer for better separation.

Commit:
Cleaned up application.html.haml and added class .no-padding to overrides.

Commit:
Reorganised css.

Commit:
Replaced current navbar with the one from the beginning.

Commit:
New way of presenting cards instead of table. In progress

Commit:
Edited navbar

Commit:
First draft of how the flash cards should look like.

Commit:
Added empy row, for separation between cards. Not very elegant.

Commit:
Deleted table, isn't needed anymore.

Commit:
Reorgansing

Commit:
Changed cholors and sizes.

Commit:
Simplified the footer

Commit:
One card is now in one 'container' row and the other rows are nested.

Commit:
Just inserted a line break.

Commit:
Product of rails generate scaffold User.

Commit:
Put the resources user to the right place

Commit:
Words renders now with _form for new. And did some css.

Commit:
You can now select both categories, semantical and syntactical

Commit:
select works lioks now better.

Commit:
Added default value nil to semantical_categories_de

Commit:
Again the footer, hopefully this time it will work

Commit:
Added a validation for uniqueness of name_de.

Commit:
Deleted some parts of the latest footer modification. It' didnt work properly.

Commit:
Now the cards optial better separated.

Commit:
Edited the show part of word.

Commit:
changed how the collection of words get displayed

Commit:
Discovered bootstrap panels, implemented that for words index

Commit:
later i will select after the categories. implemented allready a part.

Commit:
Changed col sizChanged col size

Commit:
Did a first validation uniqueness for name_de in  Word

Commit:
Deleted require_tree and require self, because the scaffold.css overrided stuff from bootstrap.

Commit:
Eddited the style of the error messsages.

Commit:
Validation gets now nice displaied. Created a shared folder for that, the other views can  render the messages from there.

Commit:
Added more validation

Commit:
Now you can go to the next and previous word in show words.

Commit:
Changed the condition statement in words show, but works at the moment.

Commit:
Commented the functions get_previous_word and get_next_word better and changed a variable name.

Commit:
Changed attribute semantical_categories_de to array
Edited word_params function in word_controller

Commit:
Edited the select in words _form

tschimbr commented on app/assets/javascripts/application.js:
```diff
@@ -18,5 +18,5 @@
 //= require nprogress-turbolinks
 //= require turboboost
-//= require gmaps
+//requre gmaps
```
Please delete if not used.

tschimbr commented on app/assets/stylesheets/application.css:
```diff
@@ -10,8 +10,7 @@
  * defined in the other CSS/SCSS files in this directory. It is generally better to create a new
  * file per style scope.
  *
- *= require_tree .
```
Don't you need to include the stylesheets for your controllers? E.g. home.css

SamJa replied:
Deleted the css which were unnecessary. They were generated by rails

tschimbr commented on app/assets/stylesheets/bootstrap_and_overrides.css.less:
```diff
@@ -97,3 +229,14 @@ table {
+
+@font-face {
+  font-family: 'Glyphicons Halflings';
+  src: url('/assets/glyphicons-halflings-regular.eot');
```
This will probably not work on production. On production each asset is compiled into a file with the hash of the content included in the filename. And hence the filename will be different that in development.

tschimbr commented on app/assets/stylesheets/home.scss:
```diff
@@ -1,3 +1,4 @@
+/*
```
Is this file (and other files like scaffolds.scss) used at all? It's not included in the CSS asset manifest file (application.css)

tschimbr commented on app/controllers/categories_controller.rb:
```diff
@@ -0,0 +1,46 @@
+    @category = Category.find(params[:id])
+
+    if @category.update_attributes(category_params)
+      flash[:notice] = 'successfully updated!'
```
Please internationalize all strings. You can use the translate helper in a controller like that:
`I18n.t('translation_key')`

SamJa replied:
Is in progress, I remove them continuously when I find some. New strings are always  internationalized.

tschimbr commented on app/controllers/words_controller.rb:
```diff
@@ -0,0 +1,57 @@
+
+    respond_to do |format|
+      if @word.save
+        format.html { redirect_to @word, notice: 'User was successfully created.' }
```
Please internationalize and adapt notice. This one is a bit misleading as it is not a user but word that has been created.

tschimbr commented on app/models/category.rb:
```diff
@@ -0,0 +1,12 @@
+  field :name_en, type: String
+  field :description_de, type: String
+  field :description_en, type: String
+  field :type_de, type: String
```
Why do we need a type for each language? The type should be the same for all languages. Or am I missing something. BTW: you cannot use the word 'type' as it is a reserved word. Use something like word_type.

tschimbr commented on app/models/user.rb:
```diff
@@ -0,0 +1,9 @@
+class User
```
Please use devise for the user and session management. See the user model in Casematch as an example: https://github.com/eonum/casematch/blob/master/app/models/user.rb

tschimbr replied:
This is still open!

tschimbr commented on app/models/word.rb:
```diff
@@ -1,7 +1,45 @@
+  field :name_en, type: String
+  field :description_de, type: String
+  field :description_en, type: String
+  field :syntactical_category_de, type: String
```
Same here: Why do we need those categories for each language? The semantical and syntactical categories should be the same in both languages.

tschimbr commented on app/models/word.rb:
```diff
@@ -1,7 +1,45 @@
+
+
+
+  def self.get_previous_word(current_word)
```
Where do you need those functions? This could get a bit slow if you have several thousands of words.

tschimbr commented on app/views/categories/edit.html.haml:
```diff
@@ -0,0 +1,15 @@
+%h2 New category
```
Please do also internationalize view texts as soon as you have created them.

tschimbr commented on app/views/layouts/application.html.haml:
```diff
@@ -42,12 +45,24 @@
                 %li= link_to 'Italiano', params.merge({ locale: :it })
                 %li= link_to 'English', params.merge({ locale: :en })
-
+    -# Content form different pages start here
```
Comments in HAML can be made with /
Hence: 
`/ Content form different pages start here`

tschimbr commented on app/views/layouts/application.html.haml:
```diff
@@ -42,12 +45,24 @@
-          = yield
\ No newline at end of file
+      = yield
+    -#Footer content
```
Same here

tschimbr commented on app/views/words/edit.html.haml:
```diff
@@ -0,0 +1,18 @@
+%h2 Edit word
```
Why don't you use the app/views/words/_form.html.haml partial here?

tschimbr commented on config/routes.rb:
```diff
@@ -1,6 +1,16 @@
 Rails.application.routes.draw do
 
+  resources :users
```
Why are users not in the locale scope? The login, registration, etc. should also be internationalized.

tschimbr commented on Gemfile.lock:
```diff
@@ -219,4 +219,4 @@ DEPENDENCIES
   uglifier (>= 1.3.0)
 
```
you can also remove geocoder, jquery-cookie-rails, mongoid-elasticsearch if you don't need it.

SamJa replied:
Done 23bcff95c35d96fac66e505562a739313e7f1d34

tschimbr commented:
@SamJa : I have commented directly in the files in the comparison view. Please ask if something is not clear.

Commit:
Added require_self and tree and deleted gmaps

Commit:
Put tree . and self back in application.css

Commit:
Deleted all css which are not in use.

Commit:
Merge branch 'semantical_categories'

Commit:
Deleted attributes which were twice.
Now every word has only one name, description etc. atrribute.

Commit:
Changed the controller because of  word model
Check commit bevore this one for more detail

Commit:
Updated all views because of changes in word model

Commit:
Insertion in semantycal_categories now possible
at the moment validation is inactive so it's easier to test things.
Had to update word_params for it in controller because of string
array parsing.

Commit:
Multiple selection for semantical_categories
Copied it from casematch, works, but not to 100%.
Have to modify it.

Commit:
Set up relations between word and category.
Semantical stuff is untested at the moment

Commit:
Changed word_params.

Commit:
Changed relationship of semantical_words

Commit:
Semantical_categories saves values in array now
Needed to add   include Mongoid::Attributes::Dynamic in word model

Commit:
Set up internationalization

Commit:
nternationalization in word controller
Notices don't flash, this is may be a problem of to fast redirecting

Commit:
nternationalization for words/_form

Commit:
Added new wordAdded new words

Commit:
changed some part in the view of words

Commit:
reorganised view

Commit:
nternationalization for views/layout

Commit:
internationalization of the footer.

Commit:
had to delete space otherwise an error shows up

Commit:
Forgot to commit that file.

Commit:
Changed from glyphicons to font awesome
and i did a small lange in the translation files.

Commit:
Forgot why i added :word_id, will keep it for the
moment

Commit:
internationalized a heading

Commit:
Reorganized locales de and en

Commit:
More internationalization for word index

Commit:
Changed the layout of words show

Commit:
added arrows in word show for navigation

Commit:
added more row classes to the view

Commit:
changed sub headings in words show

Commit:
There were some typos.

Commit:
Now the semantical categories are displayed

Commit:
Now dynamic rem and add of sem categories possible

Commit:
Implemented flex for same div size in show
Had to change the grid of the show, otherwise flex would not work.

Commit:
added smaller grid classes for mobile

tschimbr commented on app/controllers/words_controller.rb:
```diff
@@ -0,0 +1,61 @@
+  def create
+    @word = Word.new(word_params)
+    @categories = Category.all
+    puts @word
```
Do not forget to remove print statements before committing.

tschimbr replied:
remove puts

tschimbr commented on app/models/category.rb:
```diff
@@ -0,0 +1,14 @@
+  include Mongoid::Document
+
+  has_many :syntactical_words, class_name: "Word", inverse_of: :syntactical_category, :autosave => true
+  has_and_belongs_to_many :semantical_words, class_name: "Word", inverse_of: :semantical_categories, :autosave => true
```
Why has a category both semantical and syntactical words. If I recall correctly a category is either semantic or syntactic and hence its words are implicitly syntactic or semantic but never both.

tschimbr commented on app/models/word.rb:
```diff
@@ -1,7 +1,41 @@
+
+
+  # OPTIMIZE: if the word allready exist, it should ask the user if he wants to edit the existing word.
+  validates :name, uniqueness: {message: "Dieses Wort gibt es bereits." }
```
Please internationalize. You save yourself some hassle if you do this from the start and never use localized text in the code.

SamJa replied:
Done. d49d69b7410aa68f07120776cc7a14af6c2d1f9e

tschimbr commented on app/models/word.rb:
```diff
@@ -1,7 +1,41 @@
+  # those two functions are used for navigating from on word to an other.
+  # can't use the function index directly on mongo objects so i create the simple array words with all IDs to get the current index, afther that i use this index with a operatoin (+/- 1) to return the complete mongo object
+  def self.get_previous_word(current_word)
+
```
This can get very slow as we have more and more words. Consider using a linked list (store next and previous word as a link in each word).  Or you could just access a random word. Word.all does not guarantee any order. Hence what you are doing here is just obtaining a pseudo random word.
Another problem is when you call this function on the first or the last word and get an index of -1

tschimbr commented on config/locales/de.yml:
```diff
@@ -0,0 +1,64 @@
+
+  #navbar section
+  about: "Über uns"
+  category_library: "Categorien"
```
Kategorien

SamJa replied:
done. 3ba788ff48e9e9523ba501e4f7d0ff419ce2384c

tschimbr commented on config/locales/de.yml:
```diff
@@ -0,0 +1,64 @@
+
+  #footer
+  company: 'eonum AG'
+  street: 'Effingerstrasse 55'
```
Texts that cannot be localized do not need to be internationalized: "Effingerstrasse 55" is the same in all languages. '3008' as well. Though 'Bern' can be internationalized as it has different names in other languages ('Berne', 'Berna'..)

SamJa replied:
done. 6f75a5139243ff0aec682e156d89fcbaec7dd9d9 and 4ae387471aa6e4410857297e53c553bd48d35b7e

Commit:
Deleted gem wich where ununnecessary.

Commit:
Deleted code which were conected to the old gems

Commit:
Deleted lines of code from old gem

Commit:
Added new translations

Commit:
added word_warning_already_exists

Commit:
Different way to comment out

Commit:
Idea for an easier way of doing the presentation

Commit:
internationalization and new approach in show

Commit:
Changed category model to only on 'type' attribute

Commit:
Internationalization of category views

Commit:
created a from for categories, easier for changes
Also had do change the controler and did some internationalization

Commit:
made category _form working for both languages

Commit:
Created a helper MultiLanguageText
With this helper the the attributes changes automatically depending
on the language of the user

Commit:
Changed the layout of categories _index
Plus new internationalization

Commit:
changed layout of categories index.

Commit:
Translates now attribute value into diff languages
Deleted two translation in the yml files and replaced them with better ones.

Commit:
Font awesome now ind categories _index

Commit:
Edited the submit and cancle part of the views

Commit:
Will need this gem for navigate between words

Commit:
Previous way don't work any more
Because of changes category model

Commit:
Forgott to commit this earlier

Commit:
Updated  attribute names and MultiLanguageText

Commit:
Updated params and included MultiLanguageText

Commit:
Updated views because of the new attribute names

Commit:
Specified the mongoid to v5

Commit:
invere_of nil and allow blank name
Blank is necessary because otherwise there is a problem with uniqueness

Commit:
Those relationships are not necessary

Commit:
Selection in form_for works now fine

Commit:
Set a 'todo' for later

Commit:
Added gem jquery.turbolinks

Commit:
retyed button section inf words _form

Commit:
now it works not only for german words

Commit:
Nav bar fixed at top

Commit:
Typo

Commit:
Deleted t(street)

Commit:
Deleted 'zip' added 'country' instead

Commit:
Used 'word' instead of 'words' in index method

Commit:
Added css for will_pagiante

Commit:
implemented will_pagiante for words index

Commit:
Allow blanc in semantical_categories

Commit:
created js function 'goto first input field'

Commit:
Internationalization

Commit:
Deleted method which are not in use anymore

Commit:
lexical sorting for words now possible.

Commit:
Added a comment

Commit:
Deleted remove_vowels method

Commit:
Deleted next and previous from words/show

Commit:
Changed word model, new attribute language
Commit contains all follow up changes after model update

Commit:
Not needed at the moment

Commit:
Changed 'word_language' to 'language' more general

Commit:
Changed category model
Commit contains all changes which were needed after model update

Commit:
Raketask for importing words

Commit:
Changed place of storing words.csv

Commit:
Updated code which used old attribute name

Commit:
Extended rake task and created categories.csv file

Commit:
Added csv export function to words/index

Commit:
Implemented download button in words/index
Includes also the locale translations

Commit:
Add pull-right class to font awesome
Plus i changed the comment above the export button

Commit:
Delete first 'resources :users'
There was another one at the right place (below scope)

tschimbr commented on app/assets/javascripts/application.js:
```diff
@@ -11,12 +11,14 @@
-//= require gmaps
-//= require_tree .
+//= require dynamic_fields
+//= require goto_first_input_form.js
```
Those two files should already be included by require_tree

tschimbr commented on app/assets/javascripts/application.js:
```diff
@@ -11,12 +11,14 @@
+//= require dynamic_fields
+//= require goto_first_input_form.js
+//= require_self
+//= require_
```
Shouldn't this be require_tree ?

tschimbr commented on app/controllers/words_controller.rb:
```diff
@@ -0,0 +1,73 @@
+class WordsController < ApplicationController
+  require 'will_paginate/array'
```
Please put requires at the start of the file.

tschimbr commented on config/application.rb:
```diff
@@ -8,13 +8,16 @@
 require "action_view/railtie"
 require "sprockets/railtie"
 require "rails/test_unit/railtie"
+require "csv"
```
Where do you need this? Wouldn't it be better to place it directly in the file you need it.

tschimbr commented on config/initializers/string_to_sort_atoms.rb:
```diff
@@ -0,0 +1,52 @@
+class SmartSortAtom
```
Please add a short description of what this class is supposed to be doing and why we need it.

tschimbr reviewed (commented):
@SamJa The overall code quality has improved since the last review. Though there are still some small issues I have commented on. And please do not forget to document once in a while. In particular this means writing the README.md (What's the purpose of the application, how to get started, how to load data, etc..)
